### ISSUE TITLE
Systray (Patch?)

### ISSUE THREAD

**Issue by DhruvaSambrani** [kolunmi/dwlb#32]
Hi, I'd like to have a systray. Obviously, a patch may be more appropriate for such a feature, but how exactly would this work?

**Comment by kolunmi** [comment]
Hello, if you are able to implement a systray, I'll definitely merge it, no need for a patch!

Perhaps inspect the [dwm systray patch](https://dwm.suckless.org/patches/systray/dwm-systray-6.4.diff) or the source of waybar or yambar to get started.

**Comment by compromyse** [comment]
Hello, any updates?

**Comment by DhruvaSambrani** [comment]
I looked at it, it was way too hard for me

**Comment by DhruvaSambrani** [comment]
@vetu104 Thank you so much.

I do see a few issues though:
1. Systray is only displayed on the first monitor, not on the others, but it seems like the space is allocated for the status bar anyway.
2. There seems to be an additional icon, but i am not sure what that is.
![image](https://github.com/kolunmi/dwlb/assets/44899822/d752d97e-89b4-43c5-b46d-422a057e63e2)
Using [tray-control](https://github.com/andrewerf/tray-control), I don't see an entry for that at all.
3. The systray seems to overlap the status text. Is it possible to let dwlb know what the width of the status bar is, so it can render accordingly?
![image](https://github.com/kolunmi/dwlb/assets/44899822/d83eb78b-ea1e-4a29-97ab-539e94eb50f0)

**Comment by DhruvaSambrani** [comment]
4. `-nosystray` should be `-no-systray`

**Comment by DhruvaSambrani** [comment]
```patch
diff --git a/dwlb.c b/dwlb.c
index 22f224c..108010a 100644
--- a/dwlb.c
+++ b/dwlb.c
@@ -109,7 +109,7 @@
 	"	-set-top [OUTPUT]		draw bar at the top\n"	\
 	"	-set-bottom [OUTPUT]		draw bar at the bottom\n" \
 	"	-toggle-location [OUTPUT]	toggle bar location\n"	\
-	"	-nosystray			do not launch the systray program\n"	\
+	"	-no-systray			do not launch the systray program\n"	\
 	"	-traymon [OUTPUT]		set monitor name where systray will appear\n"	\
 	"Other\n"							\
 	"	-v				get version information\n" \
@@ -1431,7 +1431,7 @@ request_resize(Bar *bar, char *data)
 
 	uint32_t traywidth = (uint32_t)atoi(data);
 
-	bar->width = bar->width_orig - traywidth;
+	bar->width = bar->width_orig - buffer_scale * traywidth;
 	bar->stride = bar->width * 4;
 	bar->bufsize = bar->stride * bar->height;
 	bar->redraw = true;
@@ -1909,7 +1909,7 @@ main(int argc, char **argv)
 			if (++i >= argc)
 				DIE("Option -scale requires an argument");
 			buffer_scale = strtoul(argv[i], &argv[i] + strlen(argv[i]), 10);
-		} else if (!strcmp(argv[i], "-nosystray")) {
+		} else if (!strcmp(argv[i], "-no-systray")) {
 			systray_enabled = false;;
 		} else if (!strcmp(argv[i], "-traymon")) {
 			if (++i >= argc)
```

**Comment by vetu104** [comment]
> @vetu104 Thank you so much.
> 
> I do see a few issues though:
> 
>     1. Systray is only displayed on the first monitor, not on the others, but it seems like the space is allocated for the status bar anyway.
>        Edit: checked this out further, there seems to be no way to put a layer onto more than one monitors.
>        Edit: This is obviously only when the traymon is not set.
Yeah, this is so it works without configuring -traymon on single monitor setups.

>     2. There seems to be an additional icon, but i am not sure what that is.
2 and 3 is probably a data race. Some apps first send a broken trayitem with missing menu/missing icon while starting up, then immediately remove them and send the proper trayitem. The gdbus calls run in separate threads so I couldn't figure out how to cleanly solve this, I already tried by adding some timeout to the "unregister_statusnotifieritem" function.

**Comment by DhruvaSambrani** [comment]
Tray item: No idea. There isn't any obvious signs in the variables during register notification 

Multi-monitor: I tried to get the actual monitor that the layer is on and only send the resize request to that monitor, but I've never used gtk before. Maybe if the -traymon variable is unset, we can request it specifically to be put on the first monitor, and send the resize only to that?

**Comment by vetu104** [comment]
Copyq definitely is doing something weird. It takes 3 names on the bus, and the trayitem registration call originates from a name which doesn't actually have any data. The data is under the name registered after that.

**Comment by DhruvaSambrani** [comment]
Confirming that copyq now works

**Comment by fpf3** [comment]
This appears to have issues when interacting with `togglebar()`. Hiding the bar leaves the systray visible, and then bringing the bar back covers it up.

**Comment by fpf3** [comment]
It also appears that this breaks `-status-stdin` somehow, unless that was recently patched upstream. Unable to replicate with stock. I will investigate further.
